### DISCUSSION TITLE
How can I dynamically alter the url at runtime within links?

### DISCUSSION THREAD

**Discussion by StephaneBischoffVasco** [trpc/trpc#5574] [Q&A]
How can I dynamically alter the url at runtime within links? Or is there a better way to alter urls without recreating the client.
It does not seem possible to do. httpLink resolves its config once. 
It looks like only headers can dynamically be altered at runtime.

```
export const overridePathLink: TRPCLink<...> = () => {
  return ({ next, op }) => {
    op.url = "my new dynamic URL"; <--- Possibly change it using Feature Flag. This does not work since no url on op.
    return next(op)
  }
}

export const config: CreateTRPCClientOptions<...> = {
  transformer,
  links: [
    loggerLink(...),
    overrideUrlLink, << ----- here maybe ?
    splitLink({
      condition(op) {
        return op.context.skipBatch === true
      },
      true: httpLink({ url: `https://test.com` }), <--- Tried here, but this is static
      false: httpBatchLink({
        url: `https://test.com`,  <--- Tried here, but this is static
      }),
    }),
  ],
}

export const trpcClient = createTRPCProxyClient<...>(trpcConfig)
...

```

**Comment by KATT** [comment]
How come you'd need to change the url at runtime? Elaborate on the problem you're trying to solve

**Reply by StephaneBischoffVasco** [reply]
The problem :  Using feature flags to proxy our frontend trpc calls to different backends.

1. Our frontend imports our trpc.ts file (TRPC client created once) at different places in our codebase
2. We want to Feature Flag our backend URL path within the frontend
3. The FF value is not know on startup.  Its  received after the creation of the TRPC client (point 1)

I know there are other solutions like waiting for FF before creating the TRPC client. 
Curious to know if other solution exists.

Thanks

**Reply by KATT** [reply]
This sounds like a quite... complicated way of dealing with feature flagging, can't you do it in a way that don't creates unique deployments for each flag? 

I'm afraid we won't be able to guide you how to deal with this particular issue unless you wanna hire one of us as a consultant

**Reply by StephaneBischoffVasco** [reply]
This is temporary, so will close question as I have a short term solution.  Thanks

**Comment by jgb-solutions** [comment]
Would having a client initializer function work for you? Just call the function to create TRPC clients on the fly.

**Reply by StephaneBischoffVasco** [reply]
Yes. But its not as elegant.  I wish the links were dynamic (runtime) in that sense.

We found an alternate solution where I use 2 levels of splitLink
```
export const splitLinkBackendA= () => {
  return splitLink({
    condition(op) {
      return op.context.skipBatch === true
    },
    true: httpLink({ url: `${getBaseUrl()}/api/trpcBackendA` }),
    false: httpBatchLink({ url: `${getBaseUrl()}/api/trpcBackendA` }),
  })
}

export const splitLinkBackendB= () => {
  return splitLink({
    condition(op) {
      return op.context.skipBatch === true
    },
    true: httpLink({ url: `${getBaseUrl()}/api/trpcBackendB` }),
    false: httpBatchLink({ url: `${getBaseUrl()}/api/trpcBackendB` }),
  })
}

export const config: CreateTRPCClientOptions<...> = {
  transformer,
  links: [
    loggerLink(...),
    splitLink({
      condition(op) {
        return useBackendAFeatureFlag
      },
      true: splitLinkBackendA()
      false: splitLinkBackendB(),
    }),
  ],
}

export const trpcClient = createTRPCProxyClient<...>(trpcConfig)
...
```
